builder/utils_test.go

```diff
@@ -267,3 +268,522 @@ func TestOmitEmpty(t *testing.T) {
+				vals: []interface{}{4, 5, 6},
+				err:  nil,
+			},
+		},
+	}
+	ass := assert.New(t)
+
+	for _, v := range testData {
+		sql, val, err := BuildUpdate(v.in.table, v.in.where, v.in.update)
+		ass.Equal(v.out.cond, sql)
+		ass.Equal(v.out.vals, val)
+		ass.NoError(err)
+	}
+
+}
+
+func TestJsonArrayInsert(t *testing.T) {
+	type inStruct struct {
+		table  string
+		update map[string]interface{}
+		where  map[string]interface{}
+	}
+	type outStruct struct {
+		cond string
+		vals []interface{}
+		err  error
+	}
+	var testData = []struct {
+		in  inStruct
+		out outStruct
+	}{
+		{
+			in: inStruct{
+				table: "xx",
+				update: map[string]interface{}{
+					"a":         18,
+					"_custom_0": JsonArrayInsert("my_json", "$[0]", "first", "$[1]", 2),
+					"name":      "newName",
+				},
+				where: map[string]interface{}{
+					"_or": []map[string]interface{}{
+						{"_custom_0": JsonContains("my_json->'$[0]'", 1)},
+						{"_custom_0": JsonContains("my_json->'$[0]'", 7)},
+					},
+				},
+			},
+			out: outStruct{
+				cond: "UPDATE xx SET my_json=JSON_ARRAY_INSERT(my_json,'$[0]',?,'$[1]',?),a=?,name=? WHERE ((((? MEMBER OF(my_json->'$[0]'))) OR ((? MEMBER OF(my_json->'$[0]')))))",
+				vals: []interface{}{"first", 2, 18, "newName", 1, 7},
+				err:  nil,
+			},
+		},
+		{
+			in: inStruct{
+				table: "xx",
+				update: map[string]interface{}{
+					"_custom_0": JsonArrayInsert("my_json", "$[0]", 0, "$[0]", []int{7, 8, 9}),
+				},
+				where: map[string]interface{}{
+					"id": []int{4, 5, 6},
+				},
+			},
+			out: outStruct{
+				cond: "UPDATE xx SET my_json=JSON_ARRAY_INSERT(my_json,'$[0]',?,'$[0]',JSON_ARRAY(?,?,?)) WHERE (id IN (?,?,?))",
+				vals: []interface{}{0, 7, 8, 9, 4, 5, 6},
+				err:  nil,
+			},
+		},
+	}
+	ass := assert.New(t)
+
+	for _, v := range testData {
+		cond, val, err := BuildUpdate(v.in.table, v.in.where, v.in.update)
+		ass.Equal(v.out.cond, cond)
+		ass.Equal(v.out.vals, val)
+		ass.NoError(err)
+	}
+}
+
+func TestOrAndCustomJsonContains(t *testing.T) {
+	where := map[string]interface{}{
+		"_or1":       []map[string]interface{}{{"a": 1}, {"b": 1}},
+		"_or2":       []map[string]interface{}{{"aa": 1}, {"bb": 1}},
+		"_or3":       []map[string]interface{}{{"aaa": 1}, {"bbb": 1}},
+		"_custom_0":  JsonContains("my_json", []int(nil)),
+		"_custom_01": JsonContains("my_json", map[string]interface{}{}),
+		"_custom_1":  JsonContains("my_json->'$'", 8),
+		"_custom_3":  JsonContains("my_json->'$[last]'", map[string]interface{}{"a": 1, "b": 2}),
+		"_custom_4":  JsonContains("my_json", []int{1, 3}),
+		"_custom_5":  JsonContains("my_json->'$[0]'", nil),
+	}
+	s := "SELECT * FROM xx WHERE (JSON_CONTAINS(my_json,JSON_ARRAY()) AND JSON_CONTAINS(my_json,JSON_OBJECT()) AND (? MEMBER OF(my_json->'$')) AND JSON_CONTAINS(my_json->'$[last]',JSON_OBJECT(?,?,?,?)) AND JSON_CONTAINS(my_json,JSON_ARRAY(?,?)) AND JSON_CONTAINS(my_json->'$[0]','null') AND ((a=?) OR (b=?)) AND ((aa=?) OR (bb=?)) AND ((aaa=?) OR (bbb=?)))"
```

Comment:
Why we accept []int{nil} or empty map？Are there any use cases that people want this kind of queries?

Reply:
maybe situation
assum data {"a":[1,2]},{"a":null}.it can be use to count where a is not null
eg
```sql
 select count(*) from tb where json_contains(xx->'$.a','[]') 
```
since sql null not same with json null;  
```sql 
select * from tb where xx->'$.a' is not null; 
```
also return row {"a":null}